plugins/json_ld.ts

```diff
@@ -0,0 +1,102 @@
+import { merge } from "../core/utils/object.ts";
+import { getDataValue } from "../core/utils/data_values.ts";
+
+import type Site from "../core/site.ts";
+import type { Page } from "../core/file.ts";
+import { Graph, Thing } from "npm:schema-dts@1.1.2";
+export interface Options {
+  /** The list extensions this plugin applies to */
+  extensions?: string[];
+
+  /** The key name for the transformations definitions */
+  name?: string;
+}
+
+const defaults: Options = {
+  extensions: [".html"],
+  name: "json_ld",
+};
+export type JsonldData = Graph | Thing;
+
+// Helper function to determine if a value is an object
+function isObject(value: unknown) {
+  return value && typeof value === "object" && !Array.isArray(value);
+}
+
+/**
+ * A plugin to insert structured JSON-LD data for SEO and social media
+ * @see https://lume.land/plugins/json_ld/
+ */
+export function json_ld(userOptions?: Options) {
+  const options = merge(defaults, userOptions);
+
+  return (site: Site) => {
+    site.mergeKey(options.name, "object");
+    site.process(options.extensions, (pages) => pages.forEach(json_ld));
+
+    function json_ld(page: Page) {
+      let json_ld = page.data[options.name] as JsonldData | undefined;
+
+      if (!json_ld || !page.document) {
+        return;
+      }
+      const { document, data } = page;
+
+      // Recursive function to traverse and process JSON-LD data
+      function traverse(key: string | undefined, value: unknown): unknown {
+        if (typeof value === "string") {
+          const data_value = getDataValue(data, value);
+          // Check if the value is a URL or ID that needs to be processed
+          if (
+            (key == "@id" || key == "url") &&
+            (value.startsWith("#") || value.startsWith("/"))
```

Comment:
These aren't the only cases that can contain URLs. Take a look to the examples in this page: https://jsonld.com/article/

There are many values like `thumbnailUrl`, `contentUrl`, etc.
And some can contain an array of URLs, like `sameAs`, `target`, `image`, etc.

Right now, I'm thinking of 2 options (open for alternatives):
- Create an array of key names that can contain URLs and check if the key is in this array.
- Or force that all URLs must start with `/`. So any value starting with `/` is considered a URL to complete. For example `/about/` or `/#about`. But this solution seems a bit weak.

Reply:
I'll look into it and add it later.